File: 백준/Silver/4949. 균형잡힌 세상/균형잡힌 세상.py
while True:
  s = input()
  a = []
  if s == '.':
    break
  for i in s:
    if i == '(' or i == '[':
      a.append(i)
    elif i == ')':
      if len(a) != 0 and a[-1] == '(':
        a.pop()
      else:
        a.append(')')
        break
    elif i == ']':
      if len(a) != 0 and a[-1] == '[':
        a.pop()
      else:
        a.append(']')
        break
  if len(a) == 0:
    print('yes')
  else:
    print('no')

File: 백준/Silver/4949. 균형잡힌 세상/균형잡힌 세상.cc
#include <iostream>
#include <algorithm>
#include <vector>
#include <string>
using namespace std;

int main() {
    while (true) {
        string s;
        getline(cin, s);
        vector<char> a;
        if (s == ".") break;
        int de = 1;
        for (int i = 0; i < s.length(); i++) {
            if (a.empty() && (s[i] == '(' || s[i] == ')' || s[i] == '[' || s[i] == ']')) {
                a.push_back(s[i]);
            }
            else if (s[i] == '(' || s[i] == '[') a.push_back(s[i]);
            else if (s[i] == ')') {
                if (a.back() == '(') a.pop_back();
                else a.push_back(s[i]);
            }
            else if (s[i] == ']') {
                if (a.back() == '[') a.pop_back();
                else a.push_back(s[i]);
            }
        }
        if (a.empty()) cout << "yes \n";
        else cout << "no \n";
    }
}
